workers/unity/Packages/io.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Gdk.TransformSynchronization.CompressedQuaternion

```diff
@@ -0,0 +1,215 @@
+///     The Unity Quaternion to compress.
+/// </param>
+/// <remarks>
+///     <para>
+///         A quaternion's length must equal 1, so the largest value can be dropped and recalculated later using
+///         the "smallest three" trick. The target type is a uint, therefore 2 bits are budgeted for the index
+///         of the largest component and 10 bits per smallest component (including sign bit).
+///     </para>
+///     <para>
+///         This method is marked as unsafe due to the use of stack allocation.
+///     </para>
+/// </remarks>
+/// <returns>
+///     A uint32 representation of a quaternion.
+/// </returns>
+public static CompressedQuaternion FromUnityQuaternion(UnityEngine.Quaternion quaternion)
+{
+    // Ensure we have a unit quaternion before compression.
+    quaternion = quaternion.normalized;
+
+    // Borrow array from pool and initialise values.
+    var q = QuaternionPool.BorrowAndInitialize(quaternion);
+
+    // Iterate through quaternion to find the index of the largest component.
+    uint largestIndex = 0;
+    var largestValue = Mathf.Abs(q[largestIndex]);
+    for (uint i = 1; i < 4; ++i)
+    {
+        var componentAbsolute = Mathf.Abs(q[i]);
+        if (componentAbsolute > largestValue)
+        {
+            largestIndex = i;
+            largestValue = componentAbsolute;
+        }
+    }
+
+    // Since -q == q, transform the quaternion so that the largest component is positive. This means the sign
+    // bit of the largest component does not need to be sent.
+    uint negativeBit = quaternion[(int) largestIndex] < 0 ? 1u : 0u;
+
+    // Initialise a uint with the index of the largest component. The variable is shifted left after each
+    // section of the uint is populated. At the end of the loop, the uint has the following structure:
+    // |largest index (2)|signbit (1) + component (9)|signbit (1) + component (9)|signbit (1) + component (9)|
+    uint compressedQuaternion = largestIndex;
+    for (uint i = 0; i < 4; i++)
+    {
+        if (i != largestIndex)
+        {
+            // If quaternion needs to be transformed, flip the sign bit.
+            uint signBit = (q[i] < 0 ? 1u : 0u) ^ negativeBit;
+
+            // The maximum possible value of the second largest component in a unit quaternion is 1/sqrt(2), so
+            // translate the value from the range [0, 1/sqrt(2)] to [0, 1] for higher precision. Add 0.5f for
+            // rounding up the value before casting to a uint.
+            uint magnitude = (uint) (((1u << 9) - 1u) * (Mathf.Abs(q[i]) / SqrtHalf) + 0.5f);
+
+            // Shift uint by 10 bits then place the component's sign bit and 9-bit magnitude in the gap.
+            compressedQuaternion = (compressedQuaternion << 10) | (signBit << 9) | magnitude;
+        }
+    }
+
+    // Return array to pool.
+    QuaternionPool.ReturnArray(q);
+
+    return new CompressedQuaternion(compressedQuaternion);
+}
+
+private static class QuaternionPool
```

Comment:
I'm fine with this personally. @zeroZshadow what are your thoughts? This code could run in a multiple threads at some point in the future.

Reply:
For the record this is just a custom version of `ConcurrentPool<T>`: https://github.com/spatialos/gdk-for-unity/blob/feature/more-type-conversion-methods/workers/unity/Packages/io.improbable.gdk.core/Worker/ConnectionHandlers/MultithreadedSpatialOSConnectionHandler/ConcurrentPool.cs

Reply:
https://docs.unity3d.com/Packages/com.unity.mathematics@0.0/api/Unity.Mathematics.float4.html#Unity_Mathematics_float4_Item_System_Int32_
Maybe?

Reply:
moved to float4